sunpy/util/tests/test_scraper.py

```diff
@@ -140,18 +143,27 @@ def testURL_pattern():
     assert not s._URL_followsPattern('fd_20130410_ar_231211.fts.gz')
 
 
-@pytest.mark.xfail
-def testURL_patternMilliseconds():
+def testURL_patternMillisecondsGeneric():
     s = Scraper('fd_%Y%m%d_%H%M%S_%e.fts')
-    # NOTE: Seems that if below fails randomly - not understood why
-    #       with `== True` fails a bit less...
     assert s._URL_followsPattern('fd_20130410_231211_119.fts')
     assert not s._URL_followsPattern('fd_20130410_231211.fts.gz')
     assert not s._URL_followsPattern('fd_20130410_ar_231211.fts.gz')
 
 
+def testURL_patternMillisecondsZeroPadded():
+    # Asserts solution to ticket #1954.
+    # Milliseconds must be zero-padded in order to match URL lengths.
+    now_mock = Mock(return_value=datetime.datetime(2019, 4, 19, 0, 0, 0, 4009))
+    with patch('datetime.datetime', now=now_mock):
+        s = Scraper('fd_%Y%m%d_%H%M%S_%e.fts')
+    now_mock.assert_called_once()
+    assert s.now == 'fd_20190419_000000_004.fts'
+
+
 @pytest.mark.xfail
 def testFilesRange_sameDirectory_local():
+    # Fails due to an IsADirectoryError, wrapped in a URLError, after `requests`
+    # tries to open a directory as a binary file.
```

Comment:
So this is something we can fix or should this test actually always fail?

Reply:
Excellent question. I don't know if scanning directories was a functioning feature at some point or not. It's definitely fixable/implementable, but it'd take some effort.

Reply:
Yeah, that does seem like an undertaking. I wonder if we want to support that in the future or just drop it. What do you think @Cadair?

Reply:
I honestly have no idea. If there isn't an issue for this already we should probably open one. Also I am guessing @dpshelio wrote this test.

Reply:
There is already a check for `ftp` type urls (uris? I never know which) maybe add a check for `file` type urls and raise not implemented exception and alter the test to match? I mean it can't work as as it is?